src/caffe/net.cpp

```diff
@@ -536,9 +539,16 @@ Dtype Net<Dtype>::ForwardFromTo(int start, int end) {
   CHECK_GE(start, 0);
   CHECK_LT(end, layers_.size());
   Dtype loss = 0;
+  bool is_python;
   for (int i = start; i <= end; ++i) {
     // LOG(ERROR) << "Forwarding " << layer_names_[i];
+    is_python = strcmp(layers_[i]->type(), "Python") == 0;
+    if (is_python) {
+        scoped_gil_release.reset();
+    }
     Dtype layer_loss = layers_[i]->Forward(bottom_vecs_[i], top_vecs_[i]);
+    if (is_python)
+        scoped_gil_release = make_shared<ScopedGILRelease>();
```

Comment:
These changes are only relevant if `WITH_PYTHON_LAYER` are set.  That variable should probably be checked, so that it still works if `WITH_PYTHON_LAYER` is set to `0`.